distribution/tools/entitlement-agent/src/main/java/org/elasticsearch/entitlement/instrumentation/Instrumenter.java

```diff
@@ -0,0 +1,195 @@
+        ClassWriter writer = new ClassWriter(reader, COMPUTE_FRAMES | COMPUTE_MAXS);
+        ClassVisitor visitor = new EntitlementClassVisitor(Opcodes.ASM9, writer, className);
+        reader.accept(visitor, 0);
+        return writer.toByteArray();
+    }
+
+    class EntitlementClassVisitor extends ClassVisitor {
+        final String className;
+
+        EntitlementClassVisitor(int api, ClassVisitor classVisitor, String className) {
+            super(api, classVisitor);
+            this.className = className;
+        }
+
+        @Override
+        public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+            super.visit(version, access, name + classNameSuffix, signature, superName, interfaces);
+        }
+
+        @Override
+        public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+            var mv = super.visitMethod(access, name, descriptor, signature, exceptions);
+            boolean isStatic = (access & ACC_STATIC) != 0;
+            var voidDescriptor = Type.getMethodDescriptor(Type.VOID_TYPE, Type.getArgumentTypes(descriptor));
+            var key = new MethodKey(className, name, voidDescriptor, isStatic);
+            var instrumentationMethod = instrumentationMethods.get(key);
+            if (instrumentationMethod != null) {
+                // LOGGER.debug("Will instrument method {}", key);
+                return new EntitlementMethodVisitor(Opcodes.ASM9, mv, isStatic, descriptor, instrumentationMethod);
+            } else {
+                // LOGGER.trace("Will not instrument method {}", key);
+            }
+            return mv;
+        }
+    }
+
+    static class EntitlementMethodVisitor extends MethodVisitor {
+        private final boolean instrumentedMethodIsStatic;
+        private final String instrumentedMethodDescriptor;
+        private final Method instrumentationMethod;
+        private boolean hasCallerSensitiveAnnotation = false;
+
+        EntitlementMethodVisitor(
+            int api,
+            MethodVisitor methodVisitor,
+            boolean instrumentedMethodIsStatic,
+            String instrumentedMethodDescriptor,
+            Method instrumentationMethod
+        ) {
+            super(api, methodVisitor);
+            this.instrumentedMethodIsStatic = instrumentedMethodIsStatic;
+            this.instrumentedMethodDescriptor = instrumentedMethodDescriptor;
+            this.instrumentationMethod = instrumentationMethod;
+        }
+
+        @Override
+        public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
+            if (visible && descriptor.endsWith("CallerSensitive;")) {
+                hasCallerSensitiveAnnotation = true;
+            }
+            return super.visitAnnotation(descriptor, visible);
+        }
+
+        @Override
+        public void visitCode() {
```

Comment:
This lacks some way to protect against double transformation of the code. We should avoid that. IIRC @jdconrad had some ideas/thoughts on this.
(Maybe an annotation added to the already transformed methods, but it can also be a check of the first few instructions -- we know what they are if the method has been already transformed)

Reply:
I've added a remark about this in the Jira ticket. I'm opposed to checking for double-instrumentation.

Perhaps a good topic for the next sync?

Reply:
I think that if we do double, triple or quadruple-instrumentation is a bug. We can discuss this at the sync

Reply:
An alternative: can we "control" this in another way?
We have full control on the ES command line, since we generate that from the CLI tool. 
IF we can be sure our agent is the last to be applied, can we be reasonably sure that we are the last one to call retransform, and that will happen only once?
Just thinking out loud, I'm not sure about that -- the JDK docs say "The transformer is called when classes are loaded, when they are [redefined](https://docs.oracle.com/en/java/javase/21/docs/api/java.instrument/java/lang/instrument/Instrumentation.html#redefineClasses(java.lang.instrument.ClassDefinition...)). and [...] when they are [retransformed](https://docs.oracle.com/en/java/javase/21/docs/api/java.instrument/java/lang/instrument/Instrumentation.html#retransformClasses(java.lang.Class...))."

Likely, the double-insturmentation could happen only when retransformed -- when loaded they are fresh, when redefined they are different bytes. 
Wdyt? Can we enforce this way? Is it enough? (I still think we need a safeguard, but if I'm alone in this I will not block)

Reply:
Again, I'd like to see an actual situation where double-instrumentation actually happens. It's completely hypothetical right now.

Reply:
The risk of instrumenting twice is that we double the overhead. The risk of not instrumenting at all is a security vulnerability. To me, the latter is more worrisome than the former.

I propose we error out of this happens and cross that bridge when we come to it. Part of the white-hat hacking can be to try to coax our agent to instrument the code twice.

Reply:
I've looked more into it and I'm starting to agree that double-instrumentation is a bona fide problem. Started a slack thread [here](https://elastic.slack.com/archives/C07HQL9TNP6/p1727989548883789).

Reply:
> Again, I'd like to see an actual situation where double-instrumentation actually happens. It's completely hypothetical right now.

It's not; if we load the APM agent after this agent, it will happen. 
If we attach a profiler or debugger, it will happen.